src/metabase/query_processor/middleware/wrap_value_literals.clj

```diff
@@ -101,19 +104,129 @@
+(defmethod coerce-temporal [LocalDateTime  LocalTime] [t _target-class]  (t/local-time t))
+(defmethod coerce-temporal [OffsetDateTime LocalTime] [t _target-class]  (t/local-time t))
+(defmethod coerce-temporal [ZonedDateTime  LocalTime] [t _target-class]  (t/local-time t))
+
+(defmethod coerce-temporal [LocalDate      OffsetTime] [_t _target-class] (throw-invalid-time))
+(defmethod coerce-temporal [LocalTime      OffsetTime] [t _target-class]  (LocalTime->OffsetTime t))
+(defmethod coerce-temporal [OffsetTime     OffsetTime] [t _target-class]  t)
+(defmethod coerce-temporal [LocalDateTime  OffsetTime] [t target-class]   (coerce-temporal (t/local-time t) target-class))
+(defmethod coerce-temporal [OffsetDateTime OffsetTime] [t _target-class]  (t/offset-time t))
+(defmethod coerce-temporal [ZonedDateTime  OffsetTime] [t _target-class]  (t/offset-time t))
+
+(defn- throw-invalid-datetime []
+  (throw-invalid-conversion (i18n/tru "Invalid datetime literal: expected a date or datetime, got a time")))
+
+(defmethod coerce-temporal [LocalDate      LocalDateTime] [t _target-class]  (t/local-date-time t (t/local-time 0)))
+(defmethod coerce-temporal [LocalTime      LocalDateTime] [_t _target-class] (throw-invalid-datetime))
+(defmethod coerce-temporal [OffsetTime     LocalDateTime] [_t _target-class] (throw-invalid-datetime))
+(defmethod coerce-temporal [LocalDateTime  LocalDateTime] [t _target-class]  t)
+(defmethod coerce-temporal [OffsetDateTime LocalDateTime] [t _target-class]  (t/local-date-time t))
+(defmethod coerce-temporal [ZonedDateTime  LocalDateTime] [t _target-class]  (t/local-date-time t))
+
+(defmethod coerce-temporal [LocalDate      OffsetDateTime] [t target-class]   (coerce-temporal (t/local-date-time t (t/local-time 0)) target-class))
+(defmethod coerce-temporal [LocalTime      OffsetDateTime] [_t _target-class] (throw-invalid-datetime))
+(defmethod coerce-temporal [OffsetTime     OffsetDateTime] [_t _target-class] (throw-invalid-datetime))
+(defmethod coerce-temporal [LocalDateTime  OffsetDateTime] [t _target-class]  (t/offset-date-time t (qp.timezone/results-timezone-id)))
+(defmethod coerce-temporal [OffsetDateTime OffsetDateTime] [t _target-class]  t)
+(defmethod coerce-temporal [ZonedDateTime  OffsetDateTime] [t _target-class]  (t/offset-date-time t))
+
+(defmethod coerce-temporal [LocalDate      ZonedDateTime] [t target-class]   (coerce-temporal (t/local-date-time t (t/local-time 0)) target-class))
+(defmethod coerce-temporal [LocalTime      ZonedDateTime] [_t _target-class] (throw-invalid-datetime))
+(defmethod coerce-temporal [OffsetTime     ZonedDateTime] [_t _target-class] (throw-invalid-datetime))
+(defmethod coerce-temporal [LocalDateTime  ZonedDateTime] [t _target-class]  (t/zoned-date-time t (t/zone-id (qp.timezone/results-timezone-id))))
+(defmethod coerce-temporal [OffsetDateTime ZonedDateTime] [t _target-class]  t) ; OffsetDateTime is perfectly fine.
```

Comment:
I'm not sure it is perfectly fine, I believe adding/comparing time to an offsetdatetime across a daylight savings shift could give an incorrect result. I'm not sure when we would actually get into this conversion though to test this scenario.

```

 snoe> snoe=# set timezone = "America/Edmonton";
SET
 snoe> snoe=# select ('2024-03-09T23:59'::timestamp) at time zone 'America/Chicago' + '3 hour'::interval;
        ?column?
------------------------
 2024-03-10 01:59:00-07
(1 row)

 snoe> snoe=# select ('2024-03-09T23:59'::timestamp) at time zone 'America/Chicago' + '4 hour'::interval;
        ?column?
------------------------
 2024-03-10 03:59:00-06
(1 row)

 snoe> snoe=# set timezone = "MST";
SET
 snoe> snoe=# select ('2024-03-09T23:59'::timestamp) at time zone 'America/Chicago' + '3 hour'::interval;
        ?column?
------------------------
 2024-03-10 01:59:00-07
(1 row)

 snoe> snoe=# select ('2024-03-09T23:59'::timestamp) at time zone 'America/Chicago' + '4 hour'::interval;
        ?column?
------------------------
 2024-03-10 02:59:00-07
(1 row)

```